pkg/cgroups/manager/manager.go

```diff
@@ -0,0 +1,392 @@
+		pod:       newPod,
+		oldPod:    oldPod,
+		eventType: podUpdateEvent,
+	}
+}
+
+func (m *CgroupManager) OnDeletePod(pod *v1.Pod) {
+	if pod.Spec.NodeName != nodetypes.GetName() {
+		return
+	}
+	m.podEvents <- podEvent{
+		pod:       pod,
+		eventType: podDeleteEvent,
+	}
+}
+
+// GetPodMetadataForContainer returns pod metadata for the given container
+// cgroup id in case of success, or nil otherwise.
+func (m *CgroupManager) GetPodMetadataForContainer(cgroupId uint64) *PodMetadata {
+	if !m.enabled {
+		return nil
+	}
+	podMetaOut := make(chan *PodMetadata)
+
+	m.podEvents <- podEvent{
+		cgroupId:       cgroupId,
+		eventType:      podGetMetadataEvent,
+		podMetadataOut: podMetaOut,
+	}
+	select {
+	// We either receive pod metadata, or zero value when the channel is closed.
+	case pm := <-podMetaOut:
+		return pm
+	}
+}
+
+// Close should only be called once from daemon close.
+func (m *CgroupManager) Close() {
+	close(m.shutdown)
+}
+
+type podUID = string
+
+type podMetadata struct {
+	name       string
+	namespace  string
+	ips        []string
+	containers map[string]struct{}
+}
+
+type containerMetadata struct {
+	cgroupId    uint64
+	cgroupPath  string
+	parentPodId string
+}
+
+type podEvent struct {
+	pod            *v1.Pod
+	oldPod         *v1.Pod
+	cgroupId       uint64
+	eventType      int
+	podMetadataOut chan *PodMetadata
+}
+
+type fs interface {
+	Stat(name string) (os.FileInfo, error)
+}
+
+type cgroup interface {
+	GetCgroupID(cgroupPath string) (uint64, error)
+}
+
+type fsImpl struct{}
+
+func (f fsImpl) Stat(name string) (os.FileInfo, error) {
+	return os.Stat(name)
+}
+
+type cgroupImpl struct{}
+
+func (c cgroupImpl) GetCgroupID(cgroupPath string) (uint64, error) {
+	return cgroups.GetCgroupID(cgroupPath)
+}
+
+func initManager(fs fs, cg cgroup, channelSize int) *CgroupManager {
+	m := &CgroupManager{
+		podMetadataById:           make(map[string]*podMetadata),
+		containerMetadataByCgrpId: make(map[uint64]*containerMetadata),
+		podEvents:                 make(chan podEvent, channelSize),
+		shutdown:                  make(chan struct{}),
+	}
+	m.fschecker = fs
+	m.cgroupsChecker = cg
+	m.checkCgroupPath = new(sync.Once)
+
+	m.enable()
+	go m.processPodEvents()
+
+	return m
+}
+
+func (m *CgroupManager) enable() {
+	if !option.Config.EnableSocketLBTracing {
+		m.enabled = false
+		return
+	}
+	m.enabled = true
+	m.checkCgroupPath.Do(func() {
+		for _, path := range cgroupBasePaths {
+			if _, err := m.fschecker.Stat(path); err != nil {
+				continue
+			}
+			m.templateCgroupBasePath = path
+			break
+		}
+		if m.templateCgroupBasePath == "" {
+			log.Warnf("No valid cgroup base path found: socket " +
+				"load-balancing tracing feature will not work")
+			m.enabled = false
+		}
+	})
+
+	if m.enabled {
+		log.Info("Cgroup metadata manager is enabled")
+	}
+}
+
+func (m *CgroupManager) processPodEvents() {
+	for {
+		select {
+		case ev := <-m.podEvents:
+			if !m.enabled {
+				continue
+			}
```

Comment:
Ah I see, this is maybe a partial answer to my question above: Rather than scattering conditional checks for whether SocketLBTracing should be enabled across the codebase, this always runs a goroutine and just skips over the events if the flag is disabled?

I think the ideal model would be that we never initialize this code and never have it subscribe to the watcher events if the flags are disabled. Maybe for the short term it doesn't really matter, but it's a bit of an odd structure to have the watcher send events on this channel only for this channel to ignore the event notification.

(Practically I don't think there's likely to be that much impact with the frequency of events we're dealing with here, but I look forward to when we can refactor the k8s watchers to allow modules like this to just subscribe to events and not force the watcher to know about how to call out to these modules in order for these notifications to be processed)